src/inc/til/hash.h

```diff
@@ -0,0 +1,149 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#pragma once
+
+namespace til
+{
+    template<typename T>
+    struct hash_trait;
+
+    struct hasher
+    {
+        constexpr hasher(size_t state = FNV_offset_basis) noexcept :
+            _hash{ state } {}
+
+        template<typename T>
+        constexpr void write(const T& v) noexcept
+        {
+            hash_trait<T>{}(*this, v);
+        }
+
+        template<typename T, typename = std::enable_if_t<std::has_unique_object_representations_v<T>>>
+        constexpr void write(const T* data, size_t count) noexcept
+        {
+            write(reinterpret_cast<const uint8_t*>(data), sizeof(T) * count);
+        }
+
+        constexpr void write(const uint8_t* data, size_t count) noexcept
+        {
+            for (size_t i = 0; i < count; ++i)
+            {
+                _hash ^= static_cast<size_t>(data[i]);
+                _hash *= FNV_prime;
+            }
+        }
+
+        constexpr size_t finalize() const noexcept
+        {
+            return _hash;
+        }
+
+    private:
+#if defined(_WIN64)
+        static constexpr size_t FNV_offset_basis = 14695981039346656037ULL;
+        static constexpr size_t FNV_prime = 1099511628211ULL;
+#else
+        static constexpr size_t FNV_offset_basis = 2166136261U;
+        static constexpr size_t FNV_prime = 16777619U;
+#endif
+
+        size_t _hash = FNV_offset_basis;
+    };
+
+    namespace details
+    {
+        template<typename T, bool enable>
+        struct conditionally_enabled_hash_trait
+        {
+            constexpr void operator()(hasher& h, const T& v) const noexcept
+            {
+                h.write(reinterpret_cast<const uint8_t*>(&v), sizeof(T));
+            }
+        };
+
+        template<typename T>
+        struct conditionally_enabled_hash_trait<T, false>
+        {
+            conditionally_enabled_hash_trait() = delete;
+            conditionally_enabled_hash_trait(const conditionally_enabled_hash_trait&) = delete;
+            conditionally_enabled_hash_trait(conditionally_enabled_hash_trait&&) = delete;
+            conditionally_enabled_hash_trait& operator=(const conditionally_enabled_hash_trait&) = delete;
+            conditionally_enabled_hash_trait& operator=(conditionally_enabled_hash_trait&&) = delete;
+        };
+    }
+
+    template<typename T>
+    struct hash_trait : details::conditionally_enabled_hash_trait<T, std::has_unique_object_representations_v<T>>
+    {
+    };
+
+    template<>
+    struct hash_trait<float>
+    {
+        constexpr void operator()(hasher& h, float v) const noexcept
+        {
+            v = v == 0.0f ? 0.0f : v; // map -0 to 0

```

Comment:
Isn't this mapping +0 to +0?

Reply:
and below

Reply:
No, because for C++ those are the same thing. This code is a copy from our STL and is only active without `/fp:fast`.

Reply:
Ohhh. OK.